hera_cal/reflections.py

```diff
@@ -267,25 +239,445 @@ def write_reflections(self, output_calfits, input_calfits=None, overwrite=False,
+            flags : DataContainer, holding dfft flags (e.g. skip_wgts). Default is None.
+            side : str, options=['pos', 'neg', 'both']
+                Specifies dly_range as positive delays, negative delays or both.
+            keys : list of tuples
+                List of datacontainer baseline-pol tuples to create model for.
+            overwrite : bool
+                If dfft exists, overwrite its values.
+        """
+        # get dfft and flags
+        if dfft is None:
+            if not hasattr(self, 'dfft'):
+                raise ValueError("self.dfft doesn't exist, see self.fft_data and/or self.vis_clean")
+            dfft = self.dfft
+        if flags is None:
+            flags = DataContainer(dict([(k, np.zeros_like(dfft[k], dtype=np.bool)) for k in dfft]))
+
+        # get keys
+        if keys is None:
+            keys = dfft.keys()
+
+        # setup dictionaries
+        if not hasattr(self, 'umodes'):
+            self.umodes = DataContainer({})
+        if not hasattr(self, 'vmodes'):
+            self.vmodes = DataContainer({})
+        if not hasattr(self, 'svals'):
+            self.svals = DataContainer({})
+        if not hasattr(self, 'uflags'):
+            self.uflags = DataContainer({})
+
+        # get selection function
+        if side == 'pos':
+            select = np.where((self.delays >= dly_range[0]) & (self.delays <= dly_range[1]))[0]
+        elif side == 'neg':
+            select = np.where((self.delays >= -dly_range[1]) & (self.delays <= -dly_range[0]))[0]
+        elif side == 'both':
+            select = np.where((np.abs(self.delays) > dly_range[0]) & (np.abs(self.delays) < dly_range[1]))[0]
+
+        # iterate over keys
+        for k in keys:
+            if k in self.svals and not overwrite:
+                echo("{} exists in svals and overwrite == False, skipping...".format(k), verbose=verbose)
+                continue
+            if k not in dfft:
+                echo("{} not found in dfft, skipping...".format(k), verbose=verbose)
+                continue
+
+            # perform svd to get principal components
+            d = np.zeros_like(dfft[k])
+            d[:, select] = dfft[k][:, select]
+            u, s, v = _svd_waterfall(d)
+            Nmodes = len(s)
+
+            # append to containers
+            self.umodes[k] = u[:, :Nmodes]
+            self.vmodes[k] = v[:Nmodes, :]
+            self.svals[k] = s
+            self.uflags[k] = np.min(flags[k], axis=1, keepdims=True)
+
+        # get principal components
+        self.form_PCs(keys, overwrite=overwrite)
+
+        # append relevant metadata
+        if hasattr(dfft, 'times'):
+            self.umodes.times = dfft.times
+            self.vmodes.times = dfft.times
+            self.svals.times = dfft.times
+            self.uflags.times = dfft.times
+
+    def form_PCs(self, keys=None, u=None, v=None, overwrite=False, verbose=True):
```

Comment:
I don't understand the utility of this function. When would you want to use some other u and v than the one you found with pca_decomp but then store the result to self.pcomps. 

I'd cut most of it and fold the rest into self.pca_decomp or _form_PCs as necessary.

Reply:
There is quite a lot of utility for doing that actually, and there are two functions provided below to do exactly that. But I agree that we can fold this into `build_model` to reduce memory bloat